languages/python/oso/polar/ffi.py

```diff
@@ -183,43 +186,28 @@ def process_messages(self):
             print(self.enrich_message(msg))
 
 
-class QueryEvent:
-    def __init__(self, ptr):
-        self.ptr = ptr
-
-    def get(self):
-        return ffi.string(self.ptr).decode()
-
-    def __del__(self):
-        lib.string_free(self.ptr)
-
-
-class Error:
-    def __init__(self):
-        self.ptr = lib.polar_get_error()
-
-    def get(self, enrich_message=None):
-        return get_python_error(ffi.string(self.ptr).decode(), enrich_message)
-
-    def __del__(self):
-        lib.string_free(self.ptr)
-
-
-class Source:
-    def __init__(self, ptr):
-        self.ptr = ptr
-
-    def get(self):
-        return ffi.string(self.ptr).decode()
-
-    def __del__(self):
-        lib.string_free(self.ptr)
-
-
 def check_result(result, enrich_message=None):
-    if result == 0 or is_null(result):
-        raise Error().get(enrich_message)
-    return result
+    """ 
+    Unwrap the result by (a) extracting the pointers for
+    result and error, (b) freeing the result pointers, and then
+    (c) either returning the result pointer, or constructing and
+    raising the error.
+    """
+    # TODO: what type checking/assertions can we do here?
+
+    r = result.result
+    e = result.error
+    # This is a bit of a hack. All CResult struct types
+    # store two pointers: one to a result, one to an error
+    # result_free doesn't care about what the pointers actually
+    # point to though, so we can use the same result_free method
+    lib.result_free(ffi.cast("polar_CResult_c_void *", result))
+    if is_null(e):
+        return r
```

Comment:
Assert result is non null?

Reply:
There is a weird case where we return a null pointer back to represent none. I think it's in next message